overlord/hookstate/ctlcmd/fde_setup.go

```diff
@@ -0,0 +1,160 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ctlcmd
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/i18n"
+)
+
+type fdeSetupRequestCommand struct {
+	baseCommand
+}
+
+var shortFdeSetupRequestHelp = i18n.G("Request setup of full disk encryption")
+var longFdeSetupRequestHelp = i18n.G(`
+The fde-setup-request command is used inside the fde-setup hook. It will
+return information about what operation for full-disk encryption is
+requested and auxiliary data to complete this operation.
+
+The fde-setup hook should do what is requested and then call
+"snapctl fde-setup-result" and pass the result data to stdin.
+
+Here is an example for how the fde-setup hook is called initially:
+$ snapctl fde-setup-request
+{"op":"features"}
+$ echo '[]' | snapctl fde-setup-result
+
+Alternatively the hook could reply with:
+$ echo '{"error":"hardware-unsupported"}' | snapctl fde-setup-result
+
+And then it is called again with a request to do the initial key setup:
+$ snapctl fde-setup-request
+{"op":"initial-setup", "key": "key-to-seal", "key-name":"key-for-ubuntu-data"}
+$ echo "$sealed_key" | snapctl fde-setup-result
```

Comment:
Silly question, is there a reason we're passing raw jsons here? Given the very specific use case this is addressing it's most likely fine, but I'm just wondering about giving it more formal shape with `snapctl fde-setup-request --features=...`, `snapctl --error=...` etc with snapctl giving it proper shape. Is it because it's very open-ended and we want snapctl to be dumb it this case?

Reply:
it also makes it symmetrical,  request gives (json) data out and result takes it in.  We could also support a combo of both by allowing:

snapctl fde-setup-result <data
snapctl fde-setup-result features='[]'   (no stdin)  (but it adds complexity to detecting when to use/read stdin)

Reply:
Supporting this today is indeed a bit more work, we would have to modify `client.InternalSnapctlCmdNeedsStdin()` to look at the entire os.Args and parse them to see if it should read from stdin or not. This problem will solve itself eventually when we have real support for stdin and not the very naive implementation that we use right now.